README.md
OracleDeliveryDemo
==================

Oracle Delivery Project


data/history.txt
n0->n1	34.05267,-118.29584;34.05267,-118.29584;34.05267,-118.29477;34.05267,-118.29372;34.05267,-118.29268;34.05267,-118.29218;34.05255,-118.29166;34.05238,-118.29073;34.0524,-118.29037;34.05238,-118.28911;34.0524,-118.28775;34.0524,-118.28658;34.05238,-118.28542;34.05238,-118.28426;34.05182,-118.28426;34.05128,-118.28423;34.0508,-118.28423;34.04988,-118.28419;34.04906,-118.28418;34.04773,-118.28416;34.04727,-118.28429;34.0453,-118.28424;34.04498,-118.28423;34.04494,-118.28423;34.04375,-118.28419;34.04364,-118.2838;34.0432,-118.28284;34.04283,-118.282;34.0428,-118.28191;34.04236,-118.28098;34.04236,-118.28098;4.794416666666667-16-375-30-1.4720628815140127@NORTH,W OLYMPIC  BOULEVARD,-118.29584,34.05267,0.67,R,S HOOVER STREET,-118.28426,34.05238,0.60,L,VENICE  BOULEVARD,-118.28419,34.04375,0.21,F,S BURLINGTON AVE,-118.28098,34.04236, 0.00$0;
n0->n2	34.05267,-118.29584;34.05267,-118.29584;34.05267,-118.29477;34.05267,-118.29372;34.05267,-118.29268;34.05267,-118.29218;34.05255,-118.29166;34.05095,-118.29163;34.05075,-118.29166;34.04912,-118.29163;34.04729,-118.2916;34.04428,-118.2916;34.04266,-118.29166;34.04001,-118.29164;34.03908,-118.29163;34.03815,-118.29163;34.03746,-118.29163;34.0372,-118.29161;34.03703,-118.29163;34.03688,-118.29163;34.03671,-118.29161;34.03647,-118.29163;34.03647,-118.29105;34.03668,-118.28873;34.0367,-118.28769;34.03662,-118.28712;34.03656,-118.28622;34.03661,-118.28491;34.03676,-118.28404;34.0368,-118.28371;34.03688,-118.2831;34.03705,-118.28229;34.03712,-118.28204;34.03717,-118.28189;34.03755,-118.28036;34.03763,-118.27988;34.03764,-118.2793;34.03767,-118.27802;34.03761,-118.27727;6.021866666666667-16-715-30-2.1872455531575086@NORTH,W OLYMPIC  BOULEVARD,-118.29584,34.05267,0.24,R,S VERMONT AVE,-118.29166,34.05255,1.11,M-VLR,Merge Via Ramp onto null East,-118.29163,34.03647,0.83,R,null,-118.27802,34.03767, 0.04$0;
n0->n3	34.05267,-118.29584;34.05266,-118.29687;34.0526,-118.29789;34.0526,-118.29891;34.05261,-118.3;34.05092,-118.29999;34.04909,-118.29999;34.04724,-118.29996;34.04654,-118.30032;34.04633,-118.30032;34.04544,-118.30031;34.04457,-118.30029;34.0442,-118.30029;34.04378,-118.30032;34.04327,-118.30058;34.0428,-118.30067;34.04254,-118.3007;34.04185,-118.30077;34.0414,-118.30077;34.04086,-118.30064;34.04033,-118.30045;34.03998,-118.30029;34.03903,-118.30031;34.03812,-118.30032;34.03755,-118.30032;34.03732,-118.30032;34.03716,-118.30032;34.03697,-118.30032;34.03679,-118.30032;34.03659,-118.30032;34.03548,-118.30031;34.03548,-118.30031;4.69995-16-988-30-1.4354276075087709@WEST,W OLYMPIC  BOULEVARD,-118.29584,34.05267,0.24,L,S NORMANDIE AVE,-118.3,34.05261,1.20,L,W 23rd STREET,-118.30031,34.03548, 0.00$0;
n0->n4	34.05267,-118.29584;34.05267,-118.29584;34.05267,-118.29477;34.05267,-118.29372;34.05267,-118.29268;34.05267,-118.29218;34.05255,-118.29166;34.05238,-118.29073;34.0524,-118.29037;34.05238,-118.28911;34.0524,-118.28775;34.0524,-118.28658;34.05238,-118.28542;34.05238,-118.28426;34.05234,-118.28355;34.05222,-118.2822;34.05211,-118.28098;34.05208,-118.28072;34.0517,-118.27988;34.05118,-118.27872;34.05069,-118.27763;34.0502,-118.27654;34.04968,-118.27538;34.04915,-118.27419;34.04875,-118.27328;34.04834,-118.27238;34.04779,-118.27115;34.04732,-118.27005;34.04635,-118.27073;34.04589,-118.27106;34.04578,-118.2711;34.04561,-118.27113;34.04545,-118.2712;34.04532,-118.27128;34.04498,-118.27157;34.04443,-118.27212;34.04404,-118.27248;34.04352,-118.27288;34.04334,-118.27303;34.04298,-118.27324;34.04277,-118.27336;34.04242,-118.27357;34.04202,-118.27374;34.04161,-118.27387;34.04117,-118.27398;34.0406,-118.27406;34.03947,-118.27414;34.03923,-118.27415;34.03891,-118.27415;34.03873,-118.27419;34.03836,-118.27428;34.0383,-118.27428;34.03818,-118.27431;34.03802,-118.27434;34.03796,-118.27435;34.03792,-118.27435;34.0378,-118.27437;34.03769,-118.27438;34.03755,-118.27441;34.03702,-118.27457;34.03682,-118.27467;34.03673,-118.27477;34.03665,-118.27501;34.03667,-118.27541;34.0367,-118.27553;34.03677,-118.27563;34.03697,-118.27576;34.03717,-118.27583;34.03728,-118.27583;34.03749,-118.27573;34.03757,-118.27567;34.03769,-118.27542;34.0377,-118.27516;34.03769,-118.27438;34.03769,-118.27429;34.03769,-118.27415;34.03769,-118.27402;34.03769,-118.27374;34.03767,-118.27341;34.03751,-118.27264;34.03728,-118.27208;34.03641,-118.27077;34.03624,-118.27052;34.03604,-118.2702;34.03572,-118.26961;34.03558,-118.26933;34.03484,-118.26811;34.03442,-118.26718;34.03397,-118.26633;34.03351,-118.2655;34.03299,-118.26453;34.03244,-118.26352;34.03229,-118.26324;34.03195,-118.26244;34.03142,-118.26094;34.03111,-118.26003;34.03096,-118.25955;34.03073,-118.2588;34.03058,-118.25838;34.0303,-118.25771;34.02988,-118.25677;34.02954,-118.25623;34.02943,-118.25604;34.02908,-118.25574;34.02881,-118.25543;34.02861,-118.25511;34.02773,-118.25339;34.02765,-118.2532;34.0276,-118.25294;34.02765,-118.25272;34.02777,-118.25256;34.02794,-118.25249;34.02808,-118.2525;34.02826,-118.25262;34.02931,-118.25475;34.02992,-118.25598;34.02969,-118.25613;34.02954,-118.25623;34.02934,-118.25636;34.02905,-118.25653;34.02905,-118.25653;9.469866666666666-16-1362-30-4.339327802229603@NORTH,W OLYMPIC  BOULEVARD,-118.29584,34.05267,1.56,R,BLAINE STREET,-118.27005,34.04732,0.11,M-VLR,Merge Via Ramp onto I-10 East,-118.27106,34.04589,1.36,SR,I-10,-118.26718,34.03442,0.73,EXIT,E 17th STREET,-118.25604,34.02943, 0.19,M,E 17th STREET and turn slight left,-118.25339,34.02773,0.01,(S)--118.2532-34.02765-SL-0.08-SL,E 16th STREET,-118.25262,34.02826,0.22,L,S SAN PEDRO STREET,-118.25598,34.02992, 0.07$0;

src/edu/usc/infolab/main/Main.java
package edu.usc.infolab.main;

import ilog.concert.IloException;
import ilog.concert.IloIntVar;
import ilog.concert.IloLinearNumExpr;
import ilog.concert.IloNumVar;
import ilog.cplex.IloCplex;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

public class Main {

	public static void main(String[] args) throws IOException {
		// TODO Auto-generated method stub
		try {
			// LPTest1();
			// deliveryMIPTest();
			String dynamicDistFile = "data/dynamic_dist.txt";
			int TN = 25; // Time period number
			String distFile = "data/dist.txt";
			double[][][] staticTimeArray = DataRetrievalTest
					.readStaticDistArray(distFile, TN);
			double[][][] dynamicTimeArray = DataRetrievalTest
					.readDistArray(dynamicDistFile);

			// static road network
			System.out.println("Delivery using static route network...");
			deliveryMIPTest(staticTimeArray, dynamicTimeArray);

			// Dynamic road network
			System.out.println("Delivery using Dynamic route network...");
			deliveryMIPTest(dynamicTimeArray, dynamicTimeArray);
		} catch (IloException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	public static void deliveryMIPTest(double[][][] T,
			double[][][] dynamicTimeArray) throws IloException, IOException {
		int TN = dynamicTimeArray[0][0].length; // Time period number
		int N = 7; // node number
		double f = 15; // time period length
		double P = 50; // penalty
		double B = 1e10;
		double[] S = new double[N]; // service time
		double[] w = new double[N]; // weight associated with customer
		double[] L = new double[N], U = new double[N]; // lower and upper bound
														// of delivery time
		double U2 = TN * f;
		double maxUpperBound = 120;
		for (int i = 0; i < N; ++i) {
			S[i] = 5;
			w[i] = 0.3;
			L[i] = 0;
			U[i] = maxUpperBound;
		}
		// variables
		IloCplex cplex = new IloCplex();
		cplex.setOut(new FileOutputStream("data/log.txt"));
		IloIntVar[] X = cplex.intVarArray(N * N * TN, 0, 1);
		IloIntVar[] Yl = cplex.intVarArray(N, 0, 1);
		IloIntVar[] Yu = cplex.intVarArray(N, 0, 1);
		IloNumVar[] T2 = cplex.numVarArray(N, 0, U2); // maximum leave time
		// objective
		IloLinearNumExpr target = cplex.linearNumExpr();
		// add \sum_{i,j,t}{X_{ij} * T_{ijt}}
		for (int i = 0; i < N; ++i) {
			for (int j = 0; j < N; ++j) {
				for (int t = 0; t < TN; ++t) {
					target.addTerm(T[i][j][t], X[(i * N + j) * TN + t]);
				}
			}
		}
		// \sum_i^n{Yui + Yli} * P
		for (int i = 0; i < N; ++i) {
			target.addTerm(P, Yl[i]);
			target.addTerm(P, Yu[i]);
		}
		// \sum_i^n{w * T_i}
		for (int i = 0; i < N; ++i) {
			target.addTerm(w[i], T2[i]);
		}
		// add w[0] * Time when arrived at node 0
		for (int j = 1; j < N; ++j) {
			for (int t = 0; t < TN; ++t) {
				// target.addTerm(w[0], X[(j * N + 0) * TN + t]);
			}
		}
		cplex.addMinimize(target);

		// constraints
		// (2) T_0 = 0
		cplex.addEq(cplex.prod(1.0, T2[0]), 0, "(2)");
		// (3) \sum_{t,j} X_{ijt} = 1, (4) sum_{t,j} X_{ijt} = sum_{t,j} X_{jit}
		for (int i = 0; i < N; ++i) {
			IloLinearNumExpr exp = cplex.linearNumExpr();
			IloLinearNumExpr exp2 = cplex.linearNumExpr();
			for (int j = 0; j < N; ++j) {
				if (j == i)
					continue;
				for (int t = 0; t < TN; ++t) {
					exp.addTerm(1, X[(i * N + j) * TN + t]);
					exp2.addTerm(1, X[(j * N + i) * TN + t]);
				}
			}
			cplex.addEq(exp, 1, String.format("(3):(i=%d)", i));
			cplex.addEq(exp2, 1, String.format("(4):(i=%d)", i));

		}
		// (5) (6) (7)
		for (int t = 0; t < TN; ++t) {
			for (int i = 0; i < N; ++i) {
				for (int j = 1; j < N; ++j) {
					// if (i == j) continue;
					// (5) T_j - T_i - B X_{ijt} >= T_{ijt} + S_j - B
					IloLinearNumExpr exp = cplex.linearNumExpr();
					exp.addTerm(1, T2[j]);
					exp.addTerm(-1, T2[i]);
					exp.addTerm(-B, X[(i * N + j) * TN + t]);
					cplex.addGe(exp, T[i][j][t] + S[j] - B,
							String.format("(5):(%d, %d, %d)", i, j, t));
					// (6) T_i + B * X{ijt} <= f * (t + 1) + B
					exp = cplex.linearNumExpr();
					exp.addTerm(1, T2[i]);
					exp.addTerm(B, X[(i * N + j) * TN + t]);
					cplex.addLe(exp, f * (t + 1) + B,
							String.format("(6):(%d, %d, %d)", i, j, t));

					// (7) T_i >= f * t * X{ijt}
					exp = cplex.linearNumExpr();
					exp.addTerm(1, T2[i]);
					exp.addTerm(-f * t, X[(i * N + j) * TN + t]);
					cplex.addGe(exp, 0,
							String.format("(7):(%d, %d, %d)", i, j, t));
				}
			}

		}
		for (int i = 0; i < N; ++i) {
			// (8) T_i - B * Yu_i <= U_i, Upper bound constraint
			IloLinearNumExpr exp = cplex.linearNumExpr();
			exp.addTerm(1, T2[i]);
			exp.addTerm(-B, Yu[i]);
			cplex.addLe(exp, U[i], String.format("(8):(%d)", i));

			// (9) T_i + B * Yl_i >= L_i, Lower bound constraint
			exp = cplex.linearNumExpr();
			exp.addTerm(1, T2[i]);
			exp.addTerm(B, Yl[i]);
			cplex.addGe(exp, L[i], String.format("(9):(%d)", i));
		}
		if (cplex.solve()) {
			// Show and interpret the result
			int[] deliverySeq = new int[N];
			double[] departureTimes = new double[N];
			int[][] violation = new int[2][N];
			double objVal = cplex.getObjValue();
			cplex.output().println("Solution status = " + cplex.getStatus());
			// X[i,j,t]
			double[] val = cplex.getValues(X);
			int ncols = X.length;
			for (int col = 0; col < ncols; ++col)
				if (val[col] > 0) {
					int s = col / (N * TN);
					int e = (col - s * N * TN) / TN;
					int t = col % TN;
					deliverySeq[s] = e;
				}
			// T, departure times
			val = cplex.getValues(T2);
			for (int col = 0; col < T2.length; ++col) {
				departureTimes[col] = val[col];
			}

			val = cplex.getValues(Yl);
			for (int col = 0; col < Yl.length; ++col) {
				violation[0][col] = (int) val[col];
			}
			// bound violation
			val = cplex.getValues(Yu);
			for (int col = 0; col < Yu.length; ++col) {
				violation[1][col] = (int) val[col];
			}

			interpretResult(deliverySeq, departureTimes, violation, objVal,
					dynamicTimeArray);

			// calcuate total delivery time on dynamic road network
			// \sum_{i,j,t}{X_{ij} * T_{ijt}}
			double actualTime = 0;
			int start = 0;
			do {
				int next = deliverySeq[start];
				int t = (int) (actualTime / f);
				actualTime += dynamicTimeArray[start][next][t];
				if (next != 0) {
					actualTime += S[next]; // add service time
				}
				start = next;
				// System.out.println(actualTime);
			} while (start != 0);
			System.out.println(String.format("ActualTime: %s", actualTime));
		}
		cplex.end();
		// interpret result
	}

	private static void interpretResult(int[] deliverySeq,
			double[] departureTimes, int[][] violation, double objVal,
			double[][][] timeArray) {
		// TODO Auto-generated method stub
		// int N = deliverySeq.length;
		int cur = 0, N = deliverySeq.length;
		int f = 15;
		System.out.println("==========================================");
		System.out.println(String.format("Objective Val: %.3f", objVal));
		do {
			int t = (int) (departureTimes[cur] / f);
			int next = deliverySeq[cur];
			System.out.println(String.format("@%.3f: %d->%d",
					departureTimes[cur], cur, next));
			cur = next;
		} while (cur != 0);
		// time window violation summary
		System.out.println("Time window violation summary:");
		for (int i = 0; i < N; ++i) {
			if (violation[0][i] == 1) {
				System.out.println(String.format(
						"Lower bound of node %d is voilated.", i));
			}
			if (violation[1][i] == 1) {
				System.out.println(String.format(
						"Upper bound of node %d is voilated.", i));
			}
		}
	}

	private double getObjVal(int[] deliverySeq, double[][][] timeArray) {
		return 0;
	}
}


src/edu/usc/infolab/main/MIPTest.java
/* --------------------------------------------------------------------------
 * File: MIPex1.java
 * Version 12.6
 * --------------------------------------------------------------------------
 * Licensed Materials - Property of IBM
 * 5725-A06 5725-A29 5724-Y48 5724-Y49 5724-Y54 5724-Y55 5655-Y21
 * Copyright IBM Corporation 2001, 2013. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with
 * IBM Corp.
 * --------------------------------------------------------------------------
 *
 * MIPex1.java - Entering and optimizing a MIP problem
 */
package edu.usc.infolab.main;

import ilog.concert.*;
import ilog.cplex.*;

public class MIPTest {

	public static void main(String[] args) {
		try {
			IloCplex cplex = new IloCplex();

			IloNumVar[][] var = new IloNumVar[1][];
			IloRange[][] rng = new IloRange[1][];

			populateByRow(cplex, var, rng);

			if (cplex.solve()) {
				double[] x = cplex.getValues(var[0]);
				double[] slack = cplex.getSlacks(rng[0]);

				System.out.println("Solution status = " + cplex.getStatus());
				System.out.println("Solution value  = " + cplex.getObjValue());

				for (int j = 0; j < x.length; ++j) {
					System.out.println("Variable " + j + ": Value = " + x[j]);
				}

				for (int i = 0; i < slack.length; ++i) {
					System.out.println("Constraint " + i + ": Slack = "
							+ slack[i]);
				}
			}

			cplex.exportModel("mipex1.lp");
			cplex.end();
		} catch (IloException e) {
			System.err.println("Concert exception caught '" + e + "' caught");
		}
	}

	static void populateByRow(IloMPModeler model, IloNumVar[][] var,
			IloRange[][] rng) throws IloException {
		// First define the variables, three continuous and one integer
		double[] xlb = { 0.0, 0.0, 0.0, 2.0 };
		double[] xub = { 40.0, Double.MAX_VALUE, Double.MAX_VALUE, 3.0 };
		IloNumVarType[] xt = { IloNumVarType.Float, IloNumVarType.Float,
				IloNumVarType.Float, IloNumVarType.Int };
		IloNumVar[] x = model.numVarArray(4, xlb, xub, xt);
		var[0] = x;

		// Objective Function: maximize x0 + 2*x1 + 3*x2 + x3
		double[] objvals = { 1.0, 2.0, 3.0, 1.0 };
		model.addMaximize(model.scalProd(x, objvals));

		// Three constraints
		rng[0] = new IloRange[3];
		// - x0 + x1 + x2 + 10*x3 <= 20
		rng[0][0] = model.addLe(
				model.sum(model.prod(-1.0, x[0]), model.prod(1.0, x[1]),
						model.prod(1.0, x[2]), model.prod(10.0, x[3])), 20.0);
		// x0 - 3*x1 + x2 <= 30
		rng[0][1] = model.addLe(
				model.sum(model.prod(1.0, x[0]), model.prod(-3.0, x[1]),
						model.prod(1.0, x[2])), 30.0);
		// x1 - 3.5*x3 = 0
		rng[0][2] = model.addEq(
				model.sum(model.prod(1.0, x[1]), model.prod(-3.5, x[3])), 0.0);
	}

}


src/edu/usc/infolab/main/SimpleLPTest.java
package edu.usc.infolab.main;

import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.CharsetEncoder;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

import ilog.concert.*;
import ilog.cplex.*;

public class SimpleLPTest {

	public static void main(String[] args) throws IOException {
		// TODO Auto-generated method stub
		try {
			// LPTest1();
			deliveryMIPTest();
		} catch (IloException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	private static void LPTest2() {
		try {
			IloCplex cplex = new IloCplex();
			double[] lb = { 0.0, 0.0, 0.0 };
			double[] ub = { 40.0, Double.MAX_VALUE, Double.MAX_VALUE };
			IloNumVar[] x = cplex.numVarArray(3, lb, ub);
			double[] objvals = { 1.0, 2.0, 3.0 };
			cplex.addMaximize(cplex.scalProd(x, objvals));
			cplex.addLe(cplex.sum(cplex.prod(-1.0, x[0]),
					cplex.prod(1.0, x[1]), cplex.prod(1.0, x[2])), 20.0);
			cplex.addLe(cplex.sum(cplex.prod(1.0, x[0]),
					cplex.prod(-3.0, x[1]), cplex.prod(1.0, x[2])), 30.0);
			if (cplex.solve()) {
				cplex.output()
						.println("Solution status = " + cplex.getStatus());
				cplex.output().println(
						"Solution value = " + cplex.getObjValue());
				double[] val = cplex.getValues(x);
				int ncols = cplex.getNcols();
				for (int j = 0; j < ncols; ++j)
					cplex.output().println(
							"Column: " + j + " Value = " + val[j]);
			}
			cplex.end();
		} catch (IloException e) {
			System.err.println("Concert exception '" + e + "' caught");
		}
	}

	/**
	 * Linear Programming test
	 * 
	 * <pre>
	 * Maximize 
	 * x0 + 2x1 
	 * Subject to 
	 * x0 - x1 + 5 <= 0 
	 * x0 + x1 - 10 <= 10 
	 * Bounds 
	 * x0 >= 0 
	 * x1 >= 0 
	 * End
	 * </pre>
	 * 
	 * @throws IloException
	 */
	public static void LPTest1() throws IloException {
		IloCplex cplex = new IloCplex();
		// ArrayList<IloNumExpr> exps = new ArrayList<IloNumExpr>();
		IloNumVar[] x = cplex.numVarArray(2, 0, 1e10);
		IloLinearNumExpr exp = cplex.linearNumExpr();
		exp.addTerm(1, x[0]);
		exp.addTerm(1, x[1]);
		cplex.addLe(exp, 10, "c1");
		exp = cplex.linearNumExpr();
		exp.addTerm(1, x[0]);
		exp.addTerm(-2, x[1]);
		cplex.addGe(exp, -10, "c2");
		// cplex.addLe(cplex.sum(cplex.prod(arg0, arg1)

		// objective
		IloLinearNumExpr target = cplex.linearNumExpr();
		target.addTerm(2, x[1]);
		target.addTerm(1, x[0]);
		IloObjective obj = cplex.maximize(target);
		cplex.add(obj);
		if (cplex.solve()) {
			cplex.output().println("Solution status = " + cplex.getStatus());
			cplex.output().println("Solution value = " + cplex.getObjValue());
			double[] val = cplex.getValues(x);
			int ncols = cplex.getNcols();
			for (int j = 0; j < ncols; ++j)
				cplex.output().println("Column: " + j + " Value = " + val[j]);
		}
		cplex.end();
	}

	public static void deliveryMIPTest() throws IloException, IOException {
		double[][] T = DataRetrievalTest.getDistArray();
		int N = 6;
		int TN = 30; // Time period number
		double f = 15; // time period length
		double P = 50; // penalty
		double B = 1e10;
		double[] S = new double[N]; // service time
		double[] w = new double[N]; // weight associated with customer
		double[] U = new double[N]; // upper bound of delivery time
		double[] L = new double[N]; // lower bound of delivery time
		double U2 = 60 * f;
		double maxU = 40;
		for (int i = 0; i < N; ++i) {
			S[i] = 5;
			w[i] = 0.3;
			L[i] = 0;
			U[i] = maxU;
		}
		// variables
		IloCplex cplex = new IloCplex();
		IloIntVar[] X = cplex.intVarArray(N * N * TN, 0, 1);
		IloIntVar[] Yl = cplex.intVarArray(N, 0, 1);
		IloIntVar[] Yu = cplex.intVarArray(N, 0, 1);
		IloNumVar[] T2 = cplex.numVarArray(N, 0, U2); // leave time
		// objective
		IloLinearNumExpr target = cplex.linearNumExpr();
		// add \sum_{i,j,t}{X_{ij} * T_{ijt}}
		for (int i = 0; i < N; ++i) {
			for (int j = 0; j < N; ++j) {
				for (int t = 0; t < TN; ++t) {
					target.addTerm(T[i][j], X[(i * N + j) * N + t]);
				}
			}
		}
		// \sum_i^n{Yui + Yli} * P
		for (int i = 0; i < N; ++i) {
			target.addTerm(P, Yl[i]);
			target.addTerm(P, Yu[i]);
		}
		// \sum_i^n{w * T_i}
		for (int i = 0; i < N; ++i) {
			target.addTerm(w[i], T2[i]);
		}
		cplex.addMinimize(target);

		// constraints
		// (2) T_0 = 0
		cplex.addEq(cplex.prod(1.0, T2[0]), 0, "(2)");
		// (3) \sum_{t,j} X_{ijt} = 1, (4) sum_{t,j} X_{ijt} = sum_{t,j} X_{jit}
		for (int i = 0; i < N; ++i) {
			IloLinearNumExpr exp = cplex.linearNumExpr();
			IloLinearNumExpr exp2 = cplex.linearNumExpr();
			for (int j = 0; j < N; ++j) {
				if (j == i)
					continue;
				for (int t = 0; t < TN; ++t) {
					exp.addTerm(1, X[(i * N + j) * TN + t]);
					exp2.addTerm(1, X[(j * N + i) * TN + t]);
				}
			}
			cplex.addEq(exp, 1, String.format("(3):(i=%d)", i));
			cplex.addEq(exp2, 1, String.format("(4):(i=%d)", i));
		}
		// (5) (6) (7)
		for (int t = 0; t < TN; ++t) {
			for (int i = 0; i < N; ++i) {
				for (int j = 1; j < N; ++j) {
					// (5) T_j - T_i - B X_{ijt} >= T_{ijt} + S_j - B
					IloLinearNumExpr exp = cplex.linearNumExpr();
					exp.addTerm(1, T2[j]);
					exp.addTerm(-1, T2[i]);
					exp.addTerm(-B, X[(i * N + j) * TN + t]);
					cplex.addGe(exp, T[i][j] + S[j] - B,
							String.format("(5):(%d, %d, %d)", i, j, t));
					// (6) T_i + B * X{ijt} <= f * (t + 1) + B
					exp = cplex.linearNumExpr();
					exp.addTerm(1, T2[i]);
					exp.addTerm(B, X[(i * N + j) * TN + t]);
					cplex.addLe(exp, f * (t + 1) + B,
							String.format("(6):(%d, %d, %d)", i, j, t));

					// (7) T_i >= f * t * X{ijt}
					exp = cplex.linearNumExpr();
					exp.addTerm(1, T2[i]);
					exp.addTerm(-f * t, X[(i * N + j) * TN + t]);
					cplex.addGe(exp, 0,
							String.format("(7):(%d, %d, %d)", i, j, t));
				}
			}

		}
		for (int i = 0; i < N; ++i) {
			// (8) T_i - B * Yu_i <= U_i, Upper bound constraint
			IloLinearNumExpr exp = cplex.linearNumExpr();
			exp.addTerm(1, T2[i]);
			exp.addTerm(-B, Yu[i]);
			cplex.addLe(exp, U[i], String.format("(8):(%d)", i));

			// (9) T_i + B * Yl_i >= L_i, Lower bound constraint
			exp = cplex.linearNumExpr();
			exp.addTerm(1, T2[i]);
			exp.addTerm(B, Yl[i]);
			cplex.addGe(exp, L[i], String.format("(9):(%d)", i));
		}
		if (cplex.solve()) {

			int[] deliverySeq = new int[N];
			double[] departureTimes = new double[N];
			int[][] violation = new int[2][N];
			double objVal = cplex.getObjValue();
			// Show and interpret the result
			cplex.output().println("Solution status = " + cplex.getStatus());
			// X[i,j,t]
			double[] val = cplex.getValues(X);
			int ncols = X.length;
			for (int col = 0; col < ncols; ++col)
				if (val[col] > 0) {
					int s = col / (N * TN);
					int e = (col - s * N * TN) / TN;
					int t = col % TN;
					deliverySeq[s] = e;
					// cplex.output().println(
					// String.format("X[%d,%d,%d] = %.3f", s, e, t,
					// val[col]));
				}
			// T, departure times
			val = cplex.getValues(T2);
			for (int col = 0; col < T2.length; ++col) {
				departureTimes[col] = val[col];
				// cplex.output().println(
				// String.format("T[%d] = %.3f", col, val[col]));
			}

			val = cplex.getValues(Yl);
			for (int col = 0; col < Yl.length; ++col) {
				violation[0][col] = (int) val[col];
				// cplex.output().println(
				// String.format("Yl[%d] = %.3f", col, val[col]));
			}
			// bound violation
			val = cplex.getValues(Yu);
			for (int col = 0; col < Yu.length; ++col) {
				violation[1][col] = (int) val[col];
				// cplex.output().println(
				// String.format("Yu[%d] = %.3f", col, val[col]));
			}

			interpretResult(deliverySeq, departureTimes, violation, objVal);
		}
		cplex.end();
		// interpret result
	}

	private static void interpretResult(int[] deliverySeq,
			double[] departureTimes, int[][] violation, double objVal) {
		// TODO Auto-generated method stub
		// int N = deliverySeq.length;
		int cur = 0, N = deliverySeq.length;
		System.out.println("==========================================");
		System.out.println(String.format("Objective Val: %.3f", objVal));
		do {
			System.out.println(String.format("%d->%d: %.3f", cur,
					deliverySeq[cur], departureTimes[cur]));
			cur = deliverySeq[cur];
		} while (cur != 0);
		// time window violation summary
		System.out.println("Time window violation summary:");
		for (int i = 0; i < N; ++i) {
			if (violation[0][i] == 1) {
				System.out.println(String.format(
						"Lower bound of node %d is voilated.", i));
			}
			if (violation[1][i] == 1) {
				System.out.println(String.format(
						"Upper bound of node %d is voilated.", i));
			}
		}
	}
}
